### PR TITLE
psa_generate_key_custom

### LINKED ISSUE #9020: Flexible array members are not standard C++
Since Mbed TLS 3.6.0, one of the types declared in an Mbed TLS header is a struct with a flexible array member: [`typedef struct psa_key_production_parameters_s psa_key_production_parameters_t`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/include/psa/crypto_struct.h#L226).
```
struct psa_key_production_parameters_s {
    /* Future versions may add other fields in this structure. */
    uint32_t flags;
    uint8_t data[];
};
```

This is standard C99, but not standard C++ (any version), although many compilers accept it.

```
3.6.0 % echo '#include <psa/crypto.h>' | g++ -c -I include -std=c++2a -x c++ -
3.6.0 % echo '#include <psa/crypto.h>' | g++ -c -I include -std=c++2a -pedantic -x c++ -
In file included from include/psa/crypto.h:4828,
                 from <stdin>:1:
include/psa/crypto_struct.h:229:13: warning: ISO C++ forbids flexible array member ‘data’ [-Wpedantic]
  229 |     uint8_t data[];
      |             ^~~~
3.6.0 % echo '#include <psa/crypto.h>' | clang++ -c -I include -std=c++2a -x c++ -
3.6.0 % echo '#include <psa/crypto.h>' | clang++ -c -I include -std=c++2a -pedantic -x c++ -
In file included from <stdin>:1:
In file included from include/psa/crypto.h:4828:
include/psa/crypto_struct.h:229:13: warning: flexible array members are a C99 feature [-Wc99-extensions]
    uint8_t data[];
            ^
1 warning generated.
```

First reported [here](https://github.com/Mbed-TLS/mbedtls/issues/8866#issuecomment-2029681726).

Importance:

* Although we only advertise Mbed TLS as a C library, it is generally accepted that the headers of a C library can be consumed by a C++ compiler, and we do have a test that checks that C++ builds work (currently using GCC in non-pedantic mode, but that's originally not by design, just because we didn't think of adding `-pedantic`; currently partly by design because of [other non-strict-C++-compliance that we'd also like to fix](https://github.com/Mbed-TLS/mbedtls/issues/7087)).
* This evidently impacts many users, given that it was one of the first things reported after the 3.6 release.
* Many projects like to invoke their compiler in pedantic mode. Disabling pedantic for the sake of the Mbed TLS headers would also potentially hide other desired errors.
* Most C++ compilers do include a C99 compiler and are capable of mixing C99 and C++ features. However this is not the case for all of them. On [Godbolt](https://godbolt.org/), I could only find one compiler that doesn't accept a flexible array member with its default command line: [EDG C++](https://www.edg.com/c), which claims C++17 and C90 but not C99 support.

**Goal of this issue**: find a solution so that at least users who don't call `psa_generate_key_ext`, and so don't actually need `struct psa_key_production_parameters_s`, can compile their C++ application with a strict C++ compiler that doesn't support flexible array members. This must be done without an API change for users who are building with a C99 compiler.

Validation: `g++ -std=c++17 -pedantic` and `clang++ -std=c++17 -pedantic`, ideally on the CI, but at least manually if we haven't fixed https://github.com/Mbed-TLS/mbedtls/issues/7087.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Migrate from `psa_generate_key_ext()` and `psa_key_derivation_output_key_ext()` to `psa_generate_key_custom()` and `psa_key_derivation_output_key_custom()`:

* Define the new functions.
* Change references in the documentation to point to the new functions.
* Remove the old functions. This resolves #9020.

The API now matches the current state of https://github.com/ARM-software/psa-api/pull/194.

**Note to reviewers**: I made this pull request by rebasing https://github.com/Mbed-TLS/mbedtls/pull/9235 then adding commits to remove the old functions and fix problems that didn't arise in 3.6. I experimented a bit with writing a neater history where the existing functions and tests are modified rather than having commits to add the new stuff then commits to remove the old stuff. But this ended up with a giant commit that changes a lot of things, which wouldn't be very helpful in the history, and would be more of a pain to review.

If you'd like to rely on `git range-diff`, it's a bit overwhelmed by the amount of patches that have moved (due to files being moved under `tf-psa-crypto`, the order in which the files are listed has changed). But you do get something useful with
```
git range-diff --color-moved --creation-factor=90 mbedtls-3.6..pull/9235 development..pull/9446
```

## PR checklist

- [x] **changelog** provided
- [x] **development** here
- [x] **framework** no
- [x] **3.6** https://github.com/Mbed-TLS/mbedtls/pull/9235
- [x] **2.28 backport** N/A
- [x] **tests** provided

### CODE REVIEW

**Review by valeriosetti — APPROVED** [review]
This looks like a faithful forward-port of #9235 + the removal of "old" `_key_ext` stuff.
I would just ask @tom-cosgrove-arm to check the 2 last commits as they modify the Perl scripts and corresponding generated code for `psasim`. They look good to me, but having a second check might help :)

There seems to be a conflict with `all.sh` (due to #8226 I guess), but it should be easy to fix.

**Comment by gilles-peskine-arm** [comment]
I've rebased on top of the latest `development`. The previous version is in gilles-peskine-arm:psa_generate_key_custom-development-forward_then_remove-3. The changes to `all.sh` components are now in the corresponding components file.

```
git range-diff origin/development e92e117af8add2ca8d651802d04c442930bdb59d origin/pull/9446
```

**Review by valeriosetti — APPROVED** [review]
Thanks for the update. LGTM now :)

**Review by eleuzi01 — APPROVED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
Gatekeeper note — the [3.6 backport](https://github.com/Mbed-TLS/mbedtls/pull/9235) is waiting on #9444.

**Review by tom-cosgrove-arm — APPROVED** [review]
> I would just ask @tom-cosgrove-arm to check the 2 last commits as they modify the Perl scripts and corresponding generated code for `psasim`. They look good to me, but having a second check might help :)

They look good to me too - thanks

File: tests/scripts/components-build-system.sh [hunk 1]
```diff
@@ -65,7 +65,7 @@ component_test_cmake_out_of_source () {
     mkdir "$OUT_OF_SOURCE_DIR"
     cd "$OUT_OF_SOURCE_DIR"
     # Note: Explicitly generate files as these are turned off in releases
-    cmake -D CMAKE_BUILD_TYPE:String=Check -D GEN_FILES=ON "$MBEDTLS_ROOT_DIR"
+    cmake -D CMAKE_BUILD_TYPE:String=Check -D GEN_FILES=ON _D TEST_CPP=1 "$MBEDTLS_ROOT_DIR"
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/scripts/components-build-system.sh:68]: Oops, clearly a typo for `-D` (I do have `_` on Shift+<kbd>-</kbd> on my keyboard). I wonder what cmake is doing there.

https://github.com/Mbed-TLS/mbedtls/pull/9561

**Pull request merged**